### ISSUE TITLE
If kasa power strip connection is unavailable, it stays unavailable until reload

### ISSUE THREAD

**Issue by richievos** [reef-pi/reef-pi#1837]
This has caught me a couple times. I'm running a temp probe on a reef pi, and have it a heater and a fan on a kasa smart strip. That is great in the normal case, but when I mess with my local network and cause a connection drop out, it appears the reef-pi gets into a bad state and never tries to recover.

## Repro

I haven't fully repro'ed this step-by-step, but I believe the following is likely easy enough to trigger it:

**Setup**
1. Setup a kasa hs300 driver connected to a smart plug
2. Add a Connector->Outlet to it (or multiple)

**Break the connection**
3. Unplug/restart your wifi router, or otherwise cause the kasa to lose connectivity
4. Go into reef pi and try and do something with the kasa, such as edit/save'ing the outlet

You'll now be getting error messages. I believe about the pin being unavailable (I don't have the exact message handy). You don't actually need to play with the settings to trigger anything, it's just the easiest way to tell the system is in a bad state. Similarly if you try flipping any equipment on/off that's attached to that outlet it won't work.

At this point the smart strip won't be triggered anymore, which in my case means my temp control is borked.

Additionally, I've found at this point I start getting really weird behavior across the settings generally. Eg I feel like I've had other, unrelated, things stop working. If I try and remove/add the kasa devices, they don't recover.

Also to note, even though the pi is having connectivity issues from this point onward, the kasa itself works fine using google home or the kasa apps to control it.

**Workaround: manual**
A Configuration->Admin->Reload I believe fixes the issues. You can again edit things, and I _think_ the power control works again. I am highly confident a full reboot of the device fixes it.

**Workaround: auto**
My temporary workaround was going to be trying to setup a watchdog that'd restart if this was going on. Alternatively I was thinking of just cron job'ing a full reload every 30min. I haven't gone that far though.

## Fix
My guess is the driver just gets into a bad state and completely fails when this occurs. Given it's normal for network connectivity to blip, that's likely to be a pretty common trap that'll burn people.

As a general fallback, it might be good to trigger a full reload periodically on any driver that gets into an error state. I'm wondering if this somehow relates to #1339 in that for this specific driver it should assume/detect that the state can change outside of the pi's control, and it'd need to handle transitioning over to that state. The poller too should handle connectivity drops, and it might be the proper thing to detect a transition from `disconnected->connected` and reload itself as needed.

**Comment by ranjib** [comment]
triggering full reload will be pretty risky as it will impact rest everything. This is a risk with wifi based drivers. Whatever mitigation technique we employ, has to be contained with the driver itself.

**Comment by richievos** [comment]
This has been happening to me rather consistently, giving me a lot of concerns and almost a tank crash. Generally I'm pretty confident now it's not safe to use the wifi connection for temp controls. That somewhat seems obvious to say, but I was really hoping this would work.

I just now however have found _a_ way to reproduce this, though I'm not sure if it's what's getting me. I can't seem to cleanly reproduce this by messing with the equipment once the pi is up. Meaning if I have reef-pi running and then kill the connectivity to the sensor, I can't cause it to get in a bad state.

What does however reproduce a similar symptom is _starting_ reef pi with the connection broken. Repro steps for that below.

I'm not sure this is the _only_ way this can occur. Particularly this implies my reef pi daemon is periodically restarting. That might be happening, but I am not aware why. Next time it happens I'll check out journalctl to see if there's a correlated restart.

However, if it is being triggered by a restart, then this error likely is causative:

> May 23 13:03:43 fragtankpi reef-pi[340]: 2022/05/23 13:03:43 ERROR: Failed to initialize driver:  Frag Tank Kasa  Error: dial tcp 192.....:9999: connect: network is unreachable

## Probable cause of the issue

From digging through the code, my guess is what's happening is the `hs300, hs303` code is blowing up in `NewDriver` when the connection isn't available, causing the driver to not be loaded. Since that only happens at process start, it'd mean the connection is permanently broken.

It's probably this code:

```
func (s *HS303Strip) FetchSysInfo() error {
        // THIS RIGHT HERE WILL FAIL
	buf, err := s.command.Execute(new(Plug), true)
	if err != nil {
		return err
	}
```

That code seems to exist to fetch the info about the outlets. I looked at the other single outlet tplink/kasa integrations, and they don't do that. **That should mean this only occurs on multi-outlet tplinks, not all tplinks**. That's probably the workaround I'll do, which is just switch to using separate single outlets. I don't love that idea, but it has some nice theoretical redundancy.

## Fix
Towards fixes, I could use @ranjib your take. One fix would be to let that error continue to generate, and update the code handling drivers to periodically auto-reload any drivers that failed to load. However I'm not sure if that'd require a lot of other code to be made aware of the drivers possibly reloading on the fly. **Does the code currently support drivers reloading?**

Another fix would be to just swallow the error in NewDriver, and try and somehow have it later regenerate its list of children. That feels a little hacky if it was doing it internally, so I imagine instead this would be better done as a fix similar to the first one, but instead of reloading the full driver just reload the children.

Finally, I'm wondering as well what'd happen if we just created 6 Outlets every single time, no matter what, just with bad IDs in the event this failed. For instance, just set the id to "failed_to_load-{0}" and again later try and load the proper ID somehow.

Thoughts?

## Repro by starting with broken connection.

1. setup a driver&outlet&equipment to a kasa device
2. break the connection to the kasa device. Either unplug the device, or if running reef-pi on a dev box, turn off wifi on the dev box
3. start reef pi
4. try and toggle the equipment

You will get an error ala:

> Failed to execute temperature control logic for:Temp A F. Error:can't update outlet Frag Tank Kasa plug1, pin0 driver lookup failure: driver by id 6 not available - can't get output pin

5. Next enable connectivity (eg enable wifi on your dev box)
6. try and toggle the equipment

You'll again get that error. It's broken until you do a full reload.

**Comment by ranjib** [comment]
this is as expected. When starting up (like reload) reef-pi needs to make the network connection to load the driver, and if that fails, the entire driver, including all connectors associated with it does not work. This is different from reef-pi running continuously, and then kasa device losing connectivity. In the later case the driver is still loaded, and reef-pi is not impacted. 
if you are trying to address the first case, i.e. if network is impaired and reef-pi is reloaded leading to driver load issue then that will require some change and it will be user experience impacts. currently the outlet details are obtained upon startup

**Comment by richievos** [comment]
> this is as expected. When starting up (like reload) reef-pi needs to make the network connection to load the driver, and if that fails, the entire driver, including all connectors associated with it does not work. This is different from reef-pi running continuously, and then kasa device losing connectivity. In the later case the driver is still loaded, and reef-pi is not impacted. if you are trying to address the first case, i.e. if network is impaired and reef-pi is reloaded leading to driver load issue then that will require some change and it will be user experience impacts. currently the outlet details are obtained upon startup

While it's working as coded, I think the current implementation is not actually working equivalently with the other input/output types. The core issue is for other IO types the list of pins is static. For the kasa strip it is dynamically calculating the information at startup.

The kasa strip is effectively unrecoverable if there's a transient issue at restart, while the other IOs can recover later. The easiest example is to compare this to the behavior of the kasa plug. While that too can have issues at boot, the driver itself will come up, even if connectivity is broken. Later, when connectivity restores, the plug will auto recover but the strip will stay broken.

> then that will require some change and it will be user experience impacts. 

What user impact would there be to somehow changing this behavior? I can't think of any.

## Alternative: capture details at driver setup
Thinking about that first point, another solution is address the dynamic nature more explicitly and make it static. I don't have the code up, but I think a better solution would be to grab all the plug information at time of setting up the plug, store it, and use that going forward.

That would mean if the connectivity fails, it'd fail while you're actively working on it. It'd also mean the failure can only happen once, and not at restart when you're away from the device (eg after a power outage). It'd also work more like all the other IOs. It'd also likely be easier to implement.

Thoughts @ranjib ?

**Comment by ranjib** [comment]
@richievos I think we can just get rid of the network call all together during setup or load, and populate [details of 6  outlets](https://github.com/reef-pi/drivers/blob/main/tplink/hs300.go#L86) right in the [constructor](https://github.com/reef-pi/drivers/blob/main/tplink/hs300.go#L43) with hardcoded values of id. I think thats all we need.

**Comment by richievos** [comment]
> @richievos I think we can just get rid of the network call all together during setup or load, and populate [details of 6 outlets](https://github.com/reef-pi/drivers/blob/main/tplink/hs300.go#L86) right in the [constructor](https://github.com/reef-pi/drivers/blob/main/tplink/hs300.go#L43) with hardcoded values of id. I think thats all we need.

I actually was meaning to find something where I could see examples of that response, and play with the API directly. I originally was also thinking of just hard coding it to 6 pins, but I assumed those IDs were device specific given how it's coded. Are you saying the IDs from the device itself are pre-known?

Do you have an example of what the response from the kasa API looks like?

**Comment by ranjib** [comment]
i dont think the id is specific, but yes its a risk. But there are other risks as well, the setup function that executes the initial info gathering is also used when the new driver is added. And by design its there to enforce some sanity checks such as IP check in this case. If we skip this, users will be able to add kasa driver with wrong/unusable IP. We have to do some test connection type feature to mitigate that.
After thinking through the pros and cons, im thinking for 6.0 release at least, i dont want to pick up any work on this (lesser priority). I definitely want to improve the reliability of these wifi based drivers, but probably more comprehensively (like a thread that checks periodically for state if load or any other connection failure happen. Make the whole behavior tunable, ship with sane defaults etc. This will be bit bigger work. 

@richievos unless you plan to actively work on this actively in recent time, i;ll recommend we put this in icebucket/archive it and can reconsider if it becomes relevant (by emergence of risk or available of dev effort by any means). Feel free to  if you want to discuss further and continue the work

**Comment by richievos** [comment]
@ranjib the point regarding checking the ip address is an interesting one. Why though is that important here but not important to the other wifi outlet?

Towards delaying it, I can understand that. However I don't actually agree with closing this as completed. The underlying feature bug still exists, and as we discussed in the Reef2Reef thread I believe it's a significant bug.

When looking at Reef2Reef there's numerous other posts that are referring to the same issue occurring, or at least caused confusion in debugging. A couple I found from a quick search:

* The one we chatted on that I can't find
* https://www.reef2reef.com/threads/tp-link-hs303-driver-powercut-restore-problem-reconnecting-driver.914961/
* https://www.reef2reef.com/threads/reef-pi-an-opensource-reef-tank-controller-based-on-raspberry-pi.289256/page-919
* https://www.reef2reef.com/threads/getting-error-on-hs300-equipment-add.852628/

Honestly the other thing from searching that's giving me even more concern on this is seeing people are using these wifi strips for dosing pumps controlled by reef pi, and advocating at how great an idea it is. I thought I was alone and naive that I decided to use it for a heater, but a doser is even more extreme.

I think because usually it works, it's very easy to not realize all the ways that architecture can go wrong. The driver being there, with no disclaimers around it, I think puts a big burden on people to have knowledge of all the internal implementation details so that they make a safe decision. It feels like it's setting people up for risky things to occur.

So again, that's why I can understand saying it isn't high priority or possible to get into right now, but I personally don't agree with marking this closed completed, if only because it reduces the visibility even further.
